core/src/main/scala/org/apache/spark/rdd/ReliableRDDCheckpointData.scala

```diff
@@ -63,11 +98,20 @@ private[spark] class ReliableRDDCheckpointData[T: ClassTag](@transient private v
       throw new SparkException(s"Failed to create checkpoint path $cpDir")
     }
 
-    // Save to file, and reload it as an RDD
-    val broadcastedConf = rdd.context.broadcast(
-      new SerializableConfiguration(rdd.context.hadoopConfiguration))
-    // TODO: This is expensive because it computes the RDD again unnecessarily (SPARK-8582)
-    rdd.context.runJob(rdd, ReliableCheckpointRDD.writeCheckpointFile[T](cpDir, broadcastedConf) _)
+    val checkpointedPartitionFiles = fs.listStatus(path).map(_.getPath.getName).toSet
+    // Not all actions compute all partitions of the RDD (e.g. take). For correctness, we
+    // must checkpoint any missing partitions. TODO: avoid running another job here (SPARK-8582).
+    val missingPartitionIndices = rdd.partitions.map(_.index).filter { i =>
+      !checkpointedPartitionFiles(ReliableCheckpointRDD.checkpointFileName(i))
+    }
+    if (missingPartitionIndices.nonEmpty) {
+      // TODO: This is expensive because it computes the RDD again unnecessarily (SPARK-8582)
```

Comment:
this TODO is outdated. This is only called when the iterator is not drained, in which case it is necessary to run another job to compute the missing partitions for correctness. I would just remove this TODO.